pkg/syncer/cnsoperator/controller/csinodetopology/csinodetopology_controller.go

```diff
@@ -0,0 +1,326 @@
+	configInfo *cnsconfig.ConfigurationInfo, volumeManager volumes.Manager) error {
+	ctx, log := logger.GetNewContextWithLogger()
+	if clusterFlavor != cnstypes.CnsClusterFlavorVanilla {
+		log.Debug("Not initializing the CSINodetopology Controller as it is not a Vanilla CSI deployment")
+		return nil
+	}
+
+	coCommonInterface, err := commonco.GetContainerOrchestratorInterface(ctx, common.Kubernetes, clusterFlavor, &syncer.COInitParams)
+	if err != nil {
+		log.Errorf("failed to create CO agnostic interface. Err: %v", err)
+		return err
+	}
+	if !coCommonInterface.IsFSSEnabled(ctx, common.ImprovedVolumeTopology) {
+		log.Infof("Not initializing the CSINodetopology Controller as %s FSS is disabled", common.ImprovedVolumeTopology)
+		return nil
+	}
+
+	// Initialize kubernetes client
+	k8sclient, err := k8s.NewClient(ctx)
+	if err != nil {
+		log.Errorf("creating Kubernetes client failed. Err: %v", err)
+		return err
+	}
+	// eventBroadcaster broadcasts events on csinodetopology instances to the event sink
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartRecordingToSink(
+		&typedcorev1.EventSinkImpl{
+			Interface: k8sclient.CoreV1().Events(""),
+		},
+	)
+	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: csinodetopologyv1alpha1.GroupName})
+	return add(mgr, newReconciler(mgr, configInfo, recorder))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager, configInfo *cnsconfig.ConfigurationInfo,
+	recorder record.EventRecorder) reconcile.Reconciler {
+	return &ReconcileCSINodeTopology{client: mgr.GetClient(), scheme: mgr.GetScheme(),
+		configInfo: configInfo, recorder: recorder}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	log := logger.GetLoggerWithNoContext()
+
+	// Create a new controller
+	c, err := controller.New("csinodetopology-controller", mgr, controller.Options{Reconciler: r,
+		MaxConcurrentReconciles: defaultMaxWorkerThreadsForCSINodeTopology})
+	if err != nil {
+		log.Errorf("failed to create new CSINodetopology controller with error: %+v", err)
+		return err
+	}
+
+	// Initialize backoff duration map
+	backOffDuration = make(map[string]time.Duration)
+
+	// Predicates are used to determine under which conditions
+	// the reconcile callback will be made for an instance.
+	pred := predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			return true
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			// The CO calls NodeGetInfo API just once during the node registration,
+			// therefore we do not support updates to the spec after the CR has been reconciled.
+			return false
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			// Instances are deleted by the garbage collector automatically after the corresponding NodeVM is deleted.
+			// No reconcile operations are required.
+			return false
```

Comment:
Can we print a log message for debugging?

Reply:
Done